rmw_fastrtps_cpp/src/publisher.cpp

```diff
@@ -225,7 +232,7 @@ rmw_fastrtps_cpp::create_publisher(
   rmw_fastrtps_shared_cpp::TopicHolder topic;
   if (!rmw_fastrtps_shared_cpp::cast_or_create_topic(
       dds_participant, des_topic,
-      topic_name_mangled, type_name, topic_qos, true, &topic))
+      topic_name_mangled, type_name, topic_qos, true, &topic, info->topic_listener_))
```

Comment:
Since `cast_or_create_topic` would only create the topic the first time a specific topic name is used, this could lead to using a destroyed listener.

Consider the following situation:
- Publisher for topic `topic_a` is created => will set its `CustomTopicListener` as listener for `topic_a`
- Subscription for topic `topic_a`  is created. Its `CustomSubscriptionTopicListener` will not be used.
- Publisher is deleted => its `CustomTopicListener` is destroyed
- An inconsistent topic is detected. This will try to invoke the registered listener, which has already been destroyed

Reply:
Oh, I didn't realize that this was shared for every single topic in a process.  I'll have to make sure it always gets registered, even if the topic is already there.  Thanks for the hint.

Reply:
@MiguelCompany Could you maybe give some pointers on what we are trying to accomplish with https://github.com/ros2/rmw_fastrtps/blob/e12ff1a1be8ded9623ff770a1f4f9e4b51ca8f53/rmw_fastrtps_shared_cpp/src/utils.cpp#L56-L91 ?

In particular, it seems to do some weird things for subscribers.  For instance, if you are the first and only subscription to this topic in the system, then `desc` will be nullptr, and `is_writer_topic` will be false, so `topic_holder->topic` will remain nullptr and the function will return true.  However, if there is already a publisher *or subscriber* on the topic in this process, then `desc` will be non-null, and thus we will call `participant->create_topic()` and return true.  It seems a bit weird to me to be doing different things depending on what else has been already created in the process.

Relatedly, what is the downside to calling `participant->create_topic()` for every new subscription or publisher that comes in?

Reply:
I figured this all out, and rewrote this bit in #669.  I then rebased this on top of that, so this should be ready to go.